docs/user/testing.md

```diff
@@ -0,0 +1,151 @@
+# Testing
+
+In this section, we will talk about how to test different modules and algorithms in garage.
+
+In garage, we use [pytest](https://docs.pytest.org/en/stable/getting-started.html) to carry out tests.
+
+## Test modules
+
+All test files are stored under `garage/tests` in the garage respository.
+
+Test modules are organized in the same way as the garage main repository. Ideally, all main modules and files under `garage/src/garage` should be covered with test cases.
+
+## Prerequisites
+
+To begin testing, we suggest using the same Python manager environment i.e. `virtualenv`, `conda` as we develop on garage repository so that all packages and depencies required for testing are installed in the virutal environment.
+
+Let's begin by activating the virtual environment:
+
+```bash
+# Conda
+$ conda activate myenv
+
+# Virtualenv
+$ source myenv/bin/activate
+```
+
+Next, we need to install pytest. Generally, pytest should have already been installed upon garage installation.
+
+To get started, run pytest as follows:
+
+```bash
+#!/bin/bash
+$ pytest
+```
+
+If you haven't done so, please go ahead and check out our installation guide first. It is also recommened to check out official [pytest](https://docs.pytest.org/en/stable/getting-started.html) documentation.
+
+## Basic Pytest Usage
+
+We will start with a simple test. To be consistent with pytest requirement and garage modules, we name our test `.py` file with pre-fix `test_`.
+
+Let's write a simple test case for the vpg algorithm `garage/src/garage/tf/algos/vpg.py`.
+
+We begin by creating a file called `test_vpg.py` and put it under the tree `garage/tests/garage/tf/algos/`.
+
+We want to test VPG in the cart pole environment. Hence, we create a unitest named `test_vpg_cartpole`. Inside the function, we define the environemnt, the policy and the baselien and feed them to the VPG algorithm and run the experiment. Finally, we make sure the return value is identical to our expectation by using `assert`.
+
+Sometimes, we modele multiple tests into a class for modular and scalable structure. There is no need to subclass or anything, but make sure the prefix of the class starts with `Test`, otherwise the class will be skipped.
+
+```python
+# test_vpg.py
+
+import pytest
+... # import other dependencies
+
+class TestVPG(...):
+    def test_vpg_cartpole(self):
```

Comment:
new xunit-style tests are not encouraged. especially if a test doesn't need `setup_method` and `teardown_method`, you can just use functions which start with `test_` outside of any class.

Reply:
Got it, I added this note on the doc too.